spark/wordcount/src/main/scala/com/example/bigtable/spark/wordcount/WordCount.scala

```diff
@@ -0,0 +1,141 @@
+import org.apache.spark.SparkContext
+
+/**
+  * Basic WordCount sample of using Cloud Dataproc (managed Apache Spark)
+  * to write to Cloud Bigtable.
+  */
+object WordCount {
+  val ColumnFamily = "cf"
+  val ColumnFamilyBytes = Bytes.toBytes(ColumnFamily)
+  val ColumnNameBytes = Bytes.toBytes("Count")
+
+  def createConnection(ProjectId: String, InstanceID: String): Connection = {
+    BigtableConfiguration.connect(ProjectId, InstanceID)
+  }
+
+  /**
+    * Write the count of the word to Cloud Bigtable using the word itself
+    * as the row key.
+    *
+    * @param word    The word we are writing the count of, and row key
+    * @param count   The frequency of the word in the document
+    * @param mutator A Cloud Bigtable mutator for writing
+    */
+  def writeWordCount(word: String, count: Integer, mutator: BufferedMutator) = {
+    mutator.mutate(new Put(Bytes.toBytes(word)).
+      addColumn(ColumnFamilyBytes,
+        ColumnNameBytes,
+        Bytes.toBytes(count)))
+  }
+
+  /**
+    * Create a table in the Cloud Bigtable instance if it doesn't already
+    * exit.
+    *
+    * @param connection A Cloud Bigtable connection
+    * @param name       The table name
+    */
+  def createTableIfNotExists(connection: Connection, name: String) = {
+    val tableName = TableName.valueOf(name)
+    val admin = connection.getAdmin()
+    try {
+      if (!admin.tableExists(tableName)) {
+        val tableDescriptor = new HTableDescriptor(tableName)
+        tableDescriptor.addFamily(
+          new HColumnDescriptor(ColumnFamily))
+        admin.createTable(tableDescriptor)
+      }
+    } finally {
+      admin.close()
+    }
+  }
+
+  /**
+    * Main entry point for running the WordCount spark job and writing
+    * the results to Bigtable.
+    *
+    * @param projectId  The Google Cloud Project ID
+    * @param instanceId The Cloud Bigtable instance name
+    * @param tableName  The name of the Cloud Bigtable table to write to
+    * @param fileName   The file (local or gcs) to sort
+    * @param sc         The Spark context
+    */
+  def runner(projectId: String, instanceId: String,
+             tableName: String, fileName: String,
+             sc: SparkContext) = {
+    val createTableConnection = createConnection(projectId, instanceId)
+    try {
+      createTableIfNotExists(createTableConnection, tableName)
+    } finally {
+      createTableConnection.close()
+    }
+
+    val wordCounts = sc.textFile(fileName).
+      flatMap(_.split(" ")).
+      filter(_ != "").map((_, 1)).
+      reduceByKey((a, b) => a + b)
+
+    // Create a per-partition connection to ensure each node has a
+    // connection (partitions are on at most 1 node).
+    wordCounts.foreachPartition {
+      partition => {
+        val partitionConnection = createConnection(projectId, instanceId)
+        val table = TableName.valueOf(tableName)
+        val mutator = partitionConnection.getBufferedMutator(table)
+        try {
```

Comment:
Java and Scala have a notion of try-with-resources, which makes this type of code (mutator creation and closing) a bit easier.  [Here](https://www.phdata.io/try-with-resources-in-scala/)'s one article about it.